funsor/pattern.py

```diff
@@ -0,0 +1,36 @@
+from __future__ import absolute_import, division, print_function
+
+from unification import unify
+from unification.variable import isvar
+
+import funsor.ops as ops
+from funsor.interpreter import interpretation
+from funsor.terms import Binary, Funsor, Variable, lazy
+
+
+@lazy.register(Binary, ops.Op, Funsor, Funsor)
```

Comment:
As we discussed, it would be safest to implement this as a standalone interpretation. Even further, it would be good to prevent construction of new funsors under this interpretation:
```py
def compare(cls, *args):
    if cls is not Binary:
        raise ValueError
    op, lhs, rhs = args
    if op is ops.eq:
        return lhs is rhs
    if op is ops.neq:
        return lhs is not rhs
    raise ValueError
```